### PR TITLE
feat: EIP-5202 blueprint support with API method [APE-1117]

### PR DESCRIPTION
**Pull request by antazoey**
### What I did

Allow declaring contract types.
Defines API method and implements in `ape-ethereum`.
**Needing this for the work I am doing with CREATE and CREATE2 supportin txn trace related features**

fixes: #1132 

### How I did it

* Ecosystem can implement `encode_blueprint_contract()`
* `ape-ethereum` implements EIP-5202
* New `AccountAPI` method with default implementation for `declare()` (sends the txn)

### How to verify it

TBD, just opening early in case way off

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: src/ape/api/accounts.py [hunk 1]
```diff
@@ -238,6 +238,12 @@ def deploy(
 
         return instance
 
+    def declare(self, contract: "ContractContainer", *args, **kwargs) -> ReceiptAPI:
+        transaction = self.provider.network.ecosystem.encode_contract_blueprint(
```
Comment (fubuloubu) [inline src/ape/api/accounts.py:242]: Do we also want to add `ContractContainer.deploy_blueprint`?

File: src/ape_ethereum/ecosystem.py [hunk 2]
```diff
@@ -67,6 +67,7 @@
     "goerli": (5, 5),
     "sepolia": (11155111, 11155111),
 }
+BLUEPRINT_HEADER = HexBytes("0xfe7100")
```
Comment (fubuloubu) [inline src/ape_ethereum/ecosystem.py:70]: `0xfe71` is the header, `0x00` after that is the version byte, which is 0 to start, but is combined with a 2 bit value for the size of the length data section

**Comment by antazoey** [comment]
this isnt quite working for me yet.

```python
def main():
    import ape

    account = ape.accounts.test_accounts[0]
    contract = account.deploy(ape.project.create)
    blueprint = account.declare(ape.project.Contract)
    receipt = contract.use_blueprint(blueprint.contract_address, 3, sender=account)
    address = receipt.return_value
    new_contract = ape.project.Contract.at(address)
    tx = new_contract.foo_method(1, 2, sender=account)
    assert not tx.failed

```

**Review by fubuloubu — COMMENTED** [review]
Overall thought, would it make sense to generate some sort of intermediate object or tracked deployment of the blueprint to deploy from using `ContractContainer`?

Something like `ContractContainer.deploy(..., use_blueprint=True)`

**Comment by antazoey** [comment]
> Overall thought, would it make sense to generate some sort of intermediate object or tracked deployment of the blueprint to deploy from using `ContractContainer`?
> 
> Something like `ContractContainer.deploy(..., use_blueprint=True)`

I don't really know how to deploy from a blueprint without it happening on chain through some custom contract logic / factory stuff

File: src/ape/contracts/base.py [hunk 3]
```diff
@@ -1267,6 +1267,23 @@ def deploy(self, *args, publish: bool = False, **kwargs) -> ContractInstance:
 
         return instance
 
+    def declare(self, contract: "ContractContainer", *args, **kwargs) -> ReceiptAPI:
```
Comment (fubuloubu) [inline src/ape/contracts/base.py:1270]: `self` is `ContractContainer`?
```suggestion
    def declare(self, *args, **kwargs) -> ReceiptAPI:
```

File: src/ape_ethereum/ecosystem.py [hunk 4]
```diff
@@ -210,6 +211,22 @@ def decode_address(cls, raw_address: RawAddress) -> AddressType:
...
+        contract_bytes = (bytes_obj.to_bytes() or b"") if bytes_obj else b""
+        header = kwargs.pop("header", BLUEPRINT_HEADER)
+        blueprint_bytecode = header + HexBytes(0) + contract_bytes
+        len_bytes = len(blueprint_bytecode).to_bytes(2, "big")
+        return_data_size = kwargs.pop("return_data_size", HexBytes("0x61"))
+        return_instructions = kwargs.pop("return_instructions", HexBytes("0x3d81600a3d39f3"))
+        deploy_bytecode = HexBytes(
+            return_data_size + len_bytes + return_instructions + blueprint_bytecode
+        )
+        return self.encode_deployment(deploy_bytecode, contract_type.constructor, *args, **kwargs)
```
Comment (fubuloubu) [inline src/ape_ethereum/ecosystem.py:228]: The deployment of the blueprint shouldn't have constructor args?

File: src/ape/contracts/base.py (hunk 3 above)
Comment (antazoey) [reply src/ape/contracts/base.py]: whoops

File: tests/functional/test_accounts.py [hunk 5]
```diff
@@ -492,3 +492,8 @@ def test_using_random_mnemonic(test_accounts, temp_config):
 def test_iter_test_accounts(test_accounts):
     actual = list(iter(test_accounts))
     assert len(actual) == len(test_accounts)
+
+
+def test_declare(contract_container, sender):
+    receipt = sender.declare(contract_container)
+    assert not receipt.failed
```
Comment (fubuloubu) [inline tests/functional/test_accounts.py:499]: probably should check that `receipt.contract_address` is something?

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**